### ISSUE TITLE
FederatedSigin for Social Signin returns 'not authenticated'

### ISSUE THREAD

**Issue by jcardus** [aws-amplify/amplify-js#5133]
**Describe the bug**
Using AWS Amplify and Nuxt, I have successfully configured AWS Cognito and Google. I am using Auth.federatedSignIn(). However, after successful sign-in and once redirected to the web app the result shows the error message 'not authenticated'. But if I refresh the page it works.

**To Reproduce**
Create a new app using the latest version of Amplify and Nuxt and:

$ sudo amplify init
_? Enter a name for the project test
? Enter a name for the environment test
? Choose your default editor: IntelliJ IDEA
? Choose the type of app that you're building javascript
Please tell us about your project
? What javascript framework are you using vue
? Source Directory Path:  src
? Distribution Directory Path: dist
? Build Command:  npm run-script build
? Start Command: npm run-script serve
Using default provider  awscloudformation_

_? Do you want to use an AWS profile? Yes
? Please choose the profile you want to use default
Adding backend environment test to AWS Amplify Console app: d3qip58jparm63
⠇ Initializing project in the cloud..._

sudo amplify add auth
_Using service: Cognito, provided by: awscloudformation
 The current configured provider is Amazon Cognito. _
 
 _Do you want to use the default authentication and security configuration? Default configuration with Social Provider (Federation)
 Warning: you will not be able to edit these selections. 
 How do you want users to be able to sign in? Email
 Do you want to configure advanced settings? No, I am done.
 What domain name prefix you want us to create for you? tempb059c12f-b059c12f
 Enter your redirect signin URI: http://localhost:3000/
? Do you want to add another redirect signin URI No
 Enter your redirect signout URI: http://localhost:3000/
? Do you want to add another redirect signout URI No
 Select the social providers you want to configure for your user pool: Google_
  
_ You've opted to allow users to authenticate via Google.  If you haven't already, you'll need to go to https://developers.google.com/identity and create an App ID. _
 
_ Enter your Google Web Client ID for your OAuth flow:  xxxxxxxxxxxxxxxxxxxxx
 Enter your Google Web Client Secret for your OAuth flow:  xxxxxxxxx
Successfully added resource tempb059c12f locally_

Then this code returns 'not authenticated' after the Google login:
```

import { Auth } from 'aws-amplify';
import Amplify, * as AmplifyModules from 'aws-amplify';
import awsconfig from './aws-exports'

Amplify.configure(awsconfig);

Auth.currentAuthenticatedUser().then((data) => {
  console.log(data);
}).catch(e => {
  console.error(e);
});

```
You can clearly see on the log that the method returns before fetching the info from cognito. I've read on #4132 

> (In the meantime, we're investigating how to have Auth wait for the redirect flow's resolution before returning)

But the issue was closed, so I opened this one.

**Expected behavior**
It should return the user data.

**Screenshots**
<img width="749" alt="Screenshot 2020-03-18 at 15 06 21" src="https://user-images.githubusercontent.com/31902485/76975466-79ba0b00-692a-11ea-891e-0aad0307fd58.png">
<img width="749" alt="Screenshot 2020-03-18 at 15 06 38" src="https://user-images.githubusercontent.com/31902485/76975475-7f175580-692a-11ea-9a56-83aba0b163c5.png">
<img width="878" alt="Screenshot 2020-03-18 at 15 08 36" src="https://user-images.githubusercontent.com/31902485/76975487-83437300-692a-11ea-8412-bb166dea827e.png">

</details>

**Smartphone (please complete the following information):**

- Device: Mac
- OS: macOS Mojave
- Browser Chrome
- Version 80.0.3987.132

**Comment by jcardus** [comment]
Correction, I'm using Nuxt. I suspect it only happens with Nuxt.

**Comment by ericclemmons** [comment]
@jcardus It's not just Nuxt, it's an issue we're addressing where the user is still being logged in after redirecting back to your app.

In the meantime, you can use `Hub.listen` to keep your `user` up-to-date while the login flow completes:
> https://aws-amplify.github.io/docs/js/authentication#oauth-and-hosted-ui

**Comment by jcardus** [comment]
Thanks @ericclemmons, I was trying with AmplifyEventBus and wasn't getting any event either.

I'll try with Hub.listen.

**Comment by jcardus** [comment]
Thanks again @ericclemmons, it's working with Hub.listen, I made a specific loginRedirect page to wait for the event. This is great, I spent two days on this and wasn't finding an exit!

**Comment by ericclemmons** [comment]
@jcardus Glad to hear it's working!  If you ever find yourself stuck, feel free to reach out and let us know how we can improve.

I'm keeping this issue open because it's still an issue to fix 💪

**Comment by dabiddo** [comment]
Hello @jcardus , I'm having the same issue, can you share your nuxt code on how to listen to the Hub?

I'm following this example: https://aws-amplify.github.io/docs/js/authentication#oauth-and-hosted-ui

And I see the hub sigIn event log, but when I call the Auth.currentAuthenticatedUser, I still get the "Not signed in" message,

**Comment by ericclemmons** [comment]
@dabiddo After you refresh the page and call `Auth.currentAuthenticatedUser()`, do you correctly get the user back?

If so, then you're seeing the same race-condition behavior we're addressing: after redirecting back to your app, the user credentials are still being negotiated, so `Auth.currentAuthenticatedUser()` errors out first.  Once the user is available, `Hub` dispatches an event.

If you refresh your page after authenticated and `Auth.currentAuthenticatedUser()` continues to fail, there could be something else going on...

**Comment by jcardus** [comment]
I abandoned nuxt since it doesn't go very well with framework7.

Here's my Vue code:


```
export default {
  created() {
    Hub.listen('auth', ({ payload: { event, data }}) => {
      if (event === 'signIn') {
        this.getUser()
      }
    })
    this.getUser()
    setTimeout(this.redirect, 30000)
  },
  methods: {
    redirect() {
      const redirect = window.location.protocol + '//' + window.location.host
      this.$log.info('redirecting to', redirect)
      window.location.href = redirect
    },
    getUser: async function() {
      Auth.currentAuthenticatedUser()
        .then((data) => this.getUserData(data))
        .catch((e) => {
          this.$log.error(e)
        })
    }
  }
}

```

**Comment by elorzafe** [comment]
@jcardus

I merged a fix for this, currently is published to `unstable` tag on npm if you can try out. We will publish to `latest` soon.

**Comment by yiw679** [comment]
I'm having the same issue as well. It's not the race condition since getCurrentAuthenticatedUser returns "not authenticated" even after refreshing the page. 

For me, this social login issue only exists when I deploy my frontend to Amplify console, it works perfectly without problem on localhost. 

It's worth mentioning that even though it returns not authenticated and hub.listen won't catch any event, an external user is still added to cognito user pool.

**Comment by ericclemmons** [comment]
@yiw679 Ah!  That's helpful.  After sign-in, is the redirect working correctly at least? (You may be having similar experiences as https://github.com/aws-amplify/amplify-js/issues/5565)

**Comment by yiw679** [comment]
> @yiw679 Ah! That's helpful. After sign-in, is the redirect working correctly at least? (You may be having similar experiences as #5565)

Yeah it will redirect back to my app, but the Auth.getCurrentAuthenticatedUser call (as well as all other currentXXX methods including Hub.listen will not work). This only happens with social OAuth (Google, Facebook), Auth.SignIn() works perfectly fine. 

I have tried to do Amplify remove auth and create another new instance of auth resource, but didn't work though.

I will check out this post and leave comment here on whether it helps, thanks!

**Comment by yiw679** [comment]
@ericclemmons I came across redirecting issues before and [this](https://github.com/aws-amplify/amplify-js/issues/5127#issuecomment-524629101) solution is what I currently implemented. 

My app only has one single env, and I have tried one redirect url and multiple redirect url. As for domain, I'm using the .amplifyapp.com domain. 

The thing is that it will log in successfully on localhost, and no error information is thrown except the ("not authenticated") message makes it really hard for me to debug.
